### ISSUE TITLE
Select latest record from n-m relations

### ISSUE THREAD

**Issue by mozkarakoc** [arangodb/arangodb#2007]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.0
- [x] MacOS, version: OS X El Capitan v 10.11.6

I created a edge collection as below for user following called 'follow'
```
  {
    "_key": "1", // key attribute is auto increment
    "_id": "follow/1",
    "_from": "user/1",
    "_to": "user/2",
    "_rev": "276424388",
    "createdAt": 1471794079046
  }
```
I ensured a skiplist index on createdAt field for select latest records. Then, I create 1m user and 1m follow relation between users.

I've run this query for retrieve user followers. It returns 0.001s.
```
FOR f IN follow
FILTER f._to == "user/1"
LIMIT 10
    FOR u IN user
    FILTER u._id == f._from
RETURN u
```

I need to latest followers of user/1 and I've run these queries.

```
FOR f IN follow
FILTER f._to == "user/1"
SORT f.createdAt DESC
LIMIT 10
    FOR u IN user
    FILTER u._id == f._from
RETURN u
(2.427 s)
```

```
FOR f IN follow
FILTER f._to == "user/1"
SORT TO_NUMBER(f._key) DESC
LIMIT 10
    FOR u IN user
    FILTER u._id == f._from
RETURN u
(3.536 s)
```

Is this best practice? Can I get faster results?

**Comment by jsteemann** [comment]
I have tried the above queries on data I tried to generate according to your description.
All queries returned in around 0.001 s. But clearly the time required to execute these queries depends on the number of followers of `user/1`.

The first query that determines all followers of user 1 and then sorts them by `createdAt` should quickly find the followers using the edge index and then sort them (without any index, because the edge index was used for looking up the followers). Then after sorting it will pick the first 10 followers and find the user documents for them. Finding the users will use the primary index, which is O(1). The limiting factor in this query is the sort operation and it will slow down with the number of followers for the user.

The second query (using _key for sorting) will again quickly determine all followers using the edge index, and then sort them (without any index again). After sorting it will pick the first 10 followers and find the user documents using the primary index. The limiting factor is again the sort.

In my test data, I had used a random distribution for the followers, so given 1M users and 1M followers each user did not have many followers. Given that data, the above queries should be fast enough (< 0.001 s when I tried them). The cost of sorting should not be noticeable for a handful of documents.

If the test data is different, and user 1 has 1M followers alone, then the queries will be much slower, and the costs of the sort operation will become noticeable. So I am wondering how many followers user 1 has in your test data?

**Comment by mozkarakoc** [comment]
I'm sorry that missing description about test data. I updated description on first post. In my test data, user1 has 1M followers and I want to retrieve latest 10 followers of user 1. My queries running slower in this case. 

I tried sort by createdAt or _key. Is there a best practice for this case?

**Comment by jsteemann** [comment]
Filtering by `_to` and sorting by `_key` will only use the index on `_to` for filtering. That means the sort condition will not be satisfied by the index.

Filtering by `_to` and sorting by `createdAt` may use an index if there is a (sorted) index on `[ "_to", "createdAt" ]`. If the index is used will depend on the optimizer's decision. The optimizer will consult the selectivity estimate for the edge index on `_to` and compare it with the estimated costs for using a sorted index on `[ "_to", "createdAt" ]`. This index probably will be a skiplist index, which does not have a selectivity estimate, so the optimizer will pick some arbitrary number of expected results to compare the costs for this index. It will then pick the index with the lowest estimated filtering and sorting costs.
This may be the edge index (that cannot be used for sorting) or the other index on `[ "_to", "createdAt" ]`. Which index is used will mostly depend on the number of items present in the collection and the selectivity estimate for the edge index on `_to`.

There currently is no `FORCE INDEX` hint in AQL in order to force usage of a specific index, but if it were there then it could tell the optimizer to use a specific index if it exists.

**Comment by mozkarakoc** [comment]
I tried ensure a skiplist index on  [ `_to`, `createdAt` ] fields.  Sort condition will not be satisfied by the index again. Optimizer pick the [ `_from`, `_to` ] edge index to filtering and not used any index to sorting. 

When I added `createdAt` condition to my query, it uses createdAt skiplist index to sorting but I don't know `createdAt` value for first query. I think use a dummy for createdAt on first query but don't find any workaround timestamps. Is there any suggestion? 

What about `FORCE_INDEX` hint in AQL? Is it possible as a feature on next releases?

**Comment by jsteemann** [comment]
If you want to query for the first item, you could use a value of `0` if the `createdAt` timestamps are always greater than zero (should be if they are based on Unix timestamps).

Regarding `FORCE INDEX`, could you open a feature request issue for that? Then we can see there is demand for it and the management can decide about it. Thanks.

**Comment by mozkarakoc** [comment]
No, I want to query for latest documents in collection. So dummy is not solution in this case. I'm still worried about performance issues on retrieving the latest documents. :( 

I'll open a feature request issue but I'm wondering that is solution for this if accepted by management?

**Comment by kubz** [comment]
at this scale (1m followers) IMHO it would be better to use a key/value data (an array of X followers), the follower list change everytime someone follow a specific user, so it will be costless to store/create the follower array at this time.
most of the time, this kind of list is 99% read for 1% write, you will reduce the load on the DB as well.

A similar solution is to "cache" results of this query and regenerate it when the user will have a new follower.

maybe your use case is different ?

but I agree, FORCE INDEX would be great.

**Comment by mozkarakoc** [comment]
Thanks for your suggestion but I think we should two array for each users in key-value model. Followers and followees. First array holds which users follow to the user and second array holds the user follows to which users.  I think it is not good idea. Also with this design losing the taking advantage of graph model like friend of friend? 

On the other hand, we've a timeline in project and it's design is not possible to key-value model. Sorting on timeline collection is neccessary for us. In traditional RDMS such as MYSQL, Postgres sorting with auto increment primary key very easy. arangodb should be a solution that makes it possible.
